### ISSUE TITLE
sig files have an extra garbage line at top that crashes 'get' operation

### ISSUE THREAD

**Issue by GrantEdwards** [rickysarraf/apt-offline#146]
Starting sometime in the past week or so, something seems to have changed. Now a 'set' operation produces sig files that have an extra garbage line at the top of the file that looks something like this:

```
4 normal non-security updates
'http://us.archive.ubundu.com/....
'http://us.archive.ubundu.com/....
'http://us.archive.ubundu.com/....
'http://us.archive.ubundu.com/....
```

[The wording in that first garbage line might not be exact — I don't have an example on hand...]

That causes the 'get' operation to crash when urllib throws an exception because '4' is not a recognized protocol.

I've worked around this in my update script by using 'sed' to delete any lines in the .sig file that start with a digit, but it would be nice if that wasn't needed.

**Comment by rickysarraf** [comment]
Can you give me exact steps to reproduce this ?

**Comment by GrantEdwards** [comment]
None of my servers have any pending updates, so I can't produce it at the moment. Below is the script I was using (with the new sed operation commented out). The script is run on a machine that does have internet access and can ssh to the Ubuntu server (which does not have internet access). I tried to attache it, but it's not an allowed file type.

```
#!/bin/bash

if [ $# -ne 1 ]; then
    echo "usage: $0 <hostname>"
    exit 1
fi

RemHost=$1

rm -f $RemHost-*.zip $RemHost-$.sig

set -x
ssh -t $RemHost sudo apt-offline set --update $RemHost-update.sig
scp $RemHost:$RemHost-update.sig .

~/apt-offline/apt-offline get $RemHost-update.sig --bundle $RemHost-update.zip
scp $RemHost-update.zip $RemHost:
ssh -t $RemHost sudo apt-offline install $RemHost-update.zip
    
ssh -t $RemHost sudo apt-offline set --upgrade $RemHost-upgrade.sig
scp $RemHost:$RemHost-upgrade.sig .

set +x

# remove garbage line at top with # of upgrades
# sed -i '/^[0-9]/d' $RemHost-upgrade.sig

if [ $(stat -c '%s' $RemHost-upgrade.sig) -eq 0 ]; then
    echo
    echo -n "Empty upgrade sig file, nothing to do. Press enter: "
    read x
    set -x
else
    set -x
    ~/apt-offline/apt-offline get $RemHost-upgrade.sig --bundle $RemHost-upgrade.zip
    scp $RemHost-upgrade.zip $RemHost:
    ssh -t $RemHost sudo apt-offline install $RemHost-upgrade.zip
    ssh -t $RemHost sudo apt-get upgrade
fi

ssh -t $RemHost sudo apt-offline set --upgrade --upgrade-type dist-upgrade $RemHost-dist-upgrade.sig
scp $RemHost:$RemHost-dist-upgrade.sig .

set +x

# remove garbage line at top with # of upgrades
# sed -i '/^[0-9]/d' $RemHost-dist-upgrade.sig

if [ $(stat -c '%s' $RemHost-dist-upgrade.sig) -eq 0 ]; then
    echo
    echo -n "Empty dist-upgrade sig file, nothing to do. Press enter: "
    read x
    set -x
else
    set -x
    ~/apt-offline/apt-offline get $RemHost-dist-upgrade.sig --bundle $RemHost-dist-upgrade.zip
    scp $RemHost-dist-upgrade.zip $RemHost:
    ssh -t $RemHost sudo apt-offline install $RemHost-dist-upgrade.zip
    ssh -t $RemHost sudo apt-get dist-upgrade
fi
```

**Comment by rickysarraf** [comment]
I cannot reproduce it on my Debian systems; Bullseye as well as Sid/Unstable.

**Comment by GrantEdwards** [comment]
I was seeing it on Ubuntu server 2010.04.2 LTS. But, none of my servers have any pending updates, so at the moment, the .sig files are empty. I'll disable my "fix" and watch to see if it happens next time there are updates available.

**Comment by GrantEdwards** [comment]
Apparently this was only a temporary problem with Ubuntu Sever 20.04 LTS.  When it happened it was 100% repeatable.  Doing   `sudo apt-offline set --upgrade <filename>` generated a file with the extra line at the top every time.  I manually removed that line and updates happened normally. Today was the first time there have been available updates since then, and it's not happening now.

**Comment by GrantEdwards** [comment]
It's happening again (Ubuntu 20.04.2 LTS). I now suspect the "garbage" line only shows up when there are _security_ updates available.

```
$ apt-offline --version
1.8.2

$ sudo apt-offline set --upgrade foo.sig
Gathering details needed for 'upgrade' operation

$ cat foo.sig
8 standard security updates
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/libnss-systemd_245.4-4ubuntu3.10_amd64.deb' libnss-systemd_245.4-4ubuntu3.10_amd64.deb 96156 MD5Sum:d19e3eadd2505064cac0859fc83d6d26
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/udev_245.4-4ubuntu3.10_amd64.deb' udev_245.4-4ubuntu3.10_amd64.deb 1364816 MD5Sum:760b6933a0d45e0675ec2d2079dc1578
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/libudev1_245.4-4ubuntu3.10_amd64.deb' libudev1_245.4-4ubuntu3.10_amd64.deb 78232 MD5Sum:8ddd435b11ca865da66a8878bb98c0a0
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/systemd-sysv_245.4-4ubuntu3.10_amd64.deb' systemd-sysv_245.4-4ubuntu3.10_amd64.deb 10288 MD5Sum:c40aec17a4e060b50e5b5386fe84a7db
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/systemd-timesyncd_245.4-4ubuntu3.10_amd64.deb' systemd-timesyncd_245.4-4ubuntu3.10_amd64.deb 28096 MD5Sum:0eccb93f6d6c55c86d30bf16865b9ace
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/libpam-systemd_245.4-4ubuntu3.10_amd64.deb' libpam-systemd_245.4-4ubuntu3.10_amd64.deb 186220 MD5Sum:cf3de8c43c8592debc40c5cf08c61c85
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/systemd_245.4-4ubuntu3.10_amd64.deb' systemd_245.4-4ubuntu3.10_amd64.deb 3809616 MD5Sum:2b5bb4bc614611debfc4068146a72563
'http://us.archive.ubuntu.com/ubuntu/pool/main/s/systemd/libsystemd0_245.4-4ubuntu3.10_amd64.deb' libsystemd0_245.4-4ubuntu3.10_amd64.deb 270700 MD5Sum:f5ff9134b051256c78f3d6971ae8b1f4

```

**Comment by rickysarraf** [comment]
That is very odd. I just pulled in an Ubuntu docker image and below are my impressions.

```
root@227ac78854da:~/.../apt-offline# apt-offline  --version
1.8.2
bash: __git_ps1: command not found
15:34 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    

root@227ac78854da:~/.../apt-offline# apt-offline set --upgrade foo.sig
Gathering details needed for 'upgrade' operation
bash: __git_ps1: command not found
15:35 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    

root@227ac78854da:~/.../apt-offline# cat foo.sig 
'http://archive.ubuntu.com/ubuntu/pool/main/s/systemd/libsystemd0_245.4-4ubuntu3.10_amd64.deb' libsystemd0_245.4-4ubuntu3.10_amd64.deb 270700 MD5Sum:f5ff9134b051256c78f3d6971ae8b1f4
'http://archive.ubuntu.com/ubuntu/pool/main/s/systemd/libudev1_245.4-4ubuntu3.10_amd64.deb' libudev1_245.4-4ubuntu3.10_amd64.deb 78232 MD5Sum:8ddd435b11ca865da66a8878bb98c0a0
bash: __git_ps1: command not found
15:35 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    

root@227ac78854da:~/.../apt-offline# lsb
lsb_release  lsblk        
15:35 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
root@227ac78854da:~/.../apt-offline# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal
bash: __git_ps1: command not found
15:35 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
```

**Comment by GrantEdwards** [comment]
Are those security updates in your example?

I _think_ that on my servers the extra line at the top of the sig file only shows up when there are security updates. If there are only normal, non-security updates, then there is no extra line.

My servers are all up-to-date at the moment, but the next time there are updates, I'll check again.

**Comment by rickysarraf** [comment]
These could be either. I don't use Ubuntu so I cannot be very sure.

```
root@f510df5f8f9c:~/.../obs-api# apt policy libudev1
libudev1:
  Installed: 245.4-4ubuntu3.7
  Candidate: 245.4-4ubuntu3.10
  Version table:
     245.4-4ubuntu3.10 500
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages
 *** 245.4-4ubuntu3.7 100
        100 /var/lib/dpkg/status
     245.4-4ubuntu3 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
bash: __git_ps1: command not found
16:40 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
```

**Comment by rickysarraf** [comment]
Can you try this fix ?
https://github.com/rickysarraf/apt-offline/commit/50267a68844b70cff4c8d422df7d760a2595a65b

The content generated in the signature file is not within the control of `apt-offline`. That information is generated by `apt` and friends. So, given the distribution and the apt version, it could be different.

Thus, it makes sense for `apt-offline` to try harder to determine that the item it parses is actually a URL.

```
rrs@priyasi:~/.../apt-offline (master)$ cat /tmp/test1.uris 
8 security updates
22 normal updates
'http://deb.debian.org/debian/dists/buster-updates/contrib/dep11/icons-48x48.tar.xz' deb.debian.org_debian_dists_buster-updates_contrib_dep11_icons-48x48.tar 0 
'http://deb.debian.org/debian/dists/buster-updates/contrib/dep11/icons-64x64.tar.xz' deb.debian.org_debian_dists_buster-updates_contrib_dep11_icons-64x64.tar 0 
'http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-amd64.xz' deb.debian.org_debian_dists_buster-updates_contrib_Contents-amd64 0 
'http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-i386.xz' deb.debian.org_debian_dists_buster-updates_contrib_Contents-i386 0 
'http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-all.xz' deb.debian.org_debian_dists_buster-updates_contrib_Contents-all 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/binary-amd64/Packages.xz' deb.debian.org_debian_dists_buster-updates_non-free_binary-amd64_Packages 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/binary-i386/Packages.xz' deb.debian.org_debian_dists_buster-updates_non-free_binary-i386_Packages 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/binary-all/Packages.xz' deb.debian.org_debian_dists_buster-updates_non-free_binary-all_Packages 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en_IN.xz' deb.debian.org_debian_dists_buster-updates_non-free_i18n_Translation-en%5fIN 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en.xz' deb.debian.org_debian_dists_buster-updates_non-free_i18n_Translation-en 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en_US.xz' deb.debian.org_debian_dists_buster-updates_non-free_i18n_Translation-en%5fUS 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/Components-amd64.yml.xz' deb.debian.org_debian_dists_buster-updates_non-free_dep11_Components-amd64.yml 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/Components-all.yml.xz' deb.debian.org_debian_dists_buster-updates_non-free_dep11_Components-all.yml 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/icons-48x48.tar.xz' deb.debian.org_debian_dists_buster-updates_non-free_dep11_icons-48x48.tar 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/icons-64x64.tar.xz' deb.debian.org_debian_dists_buster-updates_non-free_dep11_icons-64x64.tar 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-amd64.xz' deb.debian.org_debian_dists_buster-updates_non-free_Contents-amd64 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-i386.xz' deb.debian.org_debian_dists_buster-updates_non-free_Contents-i386 0 
'http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-all.xz' deb.debian.org_debian_dists_buster-updates_non-free_Contents-all 0 
22:47 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@priyasi:~/.../apt-offline (master)$ ./apt-offline get /tmp/test1.uris  -d /tmp/ --proxy-host aptproxy --proxy-port 3142

Fetching APT Data

Downloading http://deb.debian.org/debian/dists/buster-updates/contrib/dep11/icons-48x48.tar.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/contrib/dep11/icons-64x64.tar.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-amd64.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-amd64.gz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-i386.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-i386.gz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/contrib/Contents-all.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/binary-amd64/Packages.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/binary-amd64/Packages.xz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/binary-i386/Packages.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/binary-i386/Packages.xz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/binary-all/Packages.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/binary-all/Packages.xz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en_IN.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en.bz2 done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/i18n/Translation-en_US.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/Components-amd64.yml.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/Components-all.yml.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/icons-48x48.tar.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/dep11/icons-64x64.tar.xz                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-amd64.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-amd64.gz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-i386.xz                                                             
http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-i386.gz done                                                             
Downloading http://deb.debian.org/debian/dists/buster-updates/non-free/Contents-all.xz                                                             
 57 /  57 items: [##############################] 100.0% of 0 KiB
Downloaded data to /tmp
22:48 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
```

**Comment by kyp44** [comment]
This commit fixed the problem for me, and I confirmed that my signature file still has the security updates line at the top. Note that I am also on Ubuntu 20.04.2 LTS for both my offline and online systems.

Thanks for the fix!

**Comment by GrantEdwards** [comment]
Going back over my notes, it looks like the first time I ran into it the problem involved only non-security updates. But then later, I saw .sig files with no extra line. Most recently, the extra line was back, but all updates were security updates. At this point, I don't know why that extra line is there sometimes but not always. I don't have any pending updates, but your fix certainly looks good to me.
